book/multi-arch-controller.md

```diff
@@ -0,0 +1,114 @@
+
+An example is shown below:
+
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  labels:
+    build.appstudio.redhat.com/multi-arch-config: hosts
+  name: host-config
+  namespace: multi-arch-controller
+data:
+  host1.address: "ec2-54-165-44-192.compute-1.amazonaws.com"
+  host1.secret: "awskeys"
+  host1.concurrency: "4"
+  host1.user: "ec2-user"
+  host1.arch: "arm64"
+  ppc1.address: "150.240.147.198"
+  ppc1.arch: "ppc64le"
+  ppc1.user: "root"
+  ppc1.secret: "awskeys"
+  ppc1.concurrency: "4"
+  ibmz1.address: "169.59.165.178"
+  ibmz1.arch: "s390x"
+  ibmz1.user: "root"
+  ibmz1.secret: "awskeys"
+  ibmz1.concurrency: "4"
+```
+
+Each host has a unique name, and provides different parameters that can be configured:
+
+- `address` The address or host name
+- `secret` The name of a secret in the `multi-arch-controller` namespace that contains the SSH keys for this host. These keys are never copied into the user namespace.
+- `concurrency` The number of builds that can be running on the host at any one time. Note that this is tracked by the controller rather than the host, so for example if you had a host referenced by both staging and production each environment is not aware of the other environments builds.
+- `user` The username to use to connect to the host
+- `arch` The hosts architecture
+
+
+### Allocation Process
+
+When a `TaskRun` is identified that meets the contract above the controller will attempt to allocate a virtual machine for them.
+
+The controller looks at the current configuration, and also looks for all running tasks with a host assigned. Using this information it picks one of the hosts at random that has the most amount of free slots. This algorithm is very simple, but works fine when all hosts are the same size. If we are mixing hosts of different sizes then we may need a better algorithm.
+
+Once the host is selected the controller runs a 'provisioning' task to prepare the host. This task currently does the following:
+
+- Create a new user specifically for this build
+- Generate a new pair of SSH keys for this user and install them
+- Copy the private key for this user back to the cluster and create the required secret with this information
+
+Once the secret has been created then the original `TaskRun` will start, and can use the new user on the remote host to perform the build. This approach means that every build is run under a different user, and only low privilege ephemeral SSH keys end up in the user's namespace.
+
+Which hosts are allocated with which tasks are tracked by adding additional labels onto the `TaskRun` object.
+
+If the provisioning task fails then a label is added to the `TaskRun` indicating that it has failed and allocation is re-tried, excluding all hosts in the failed list. If all of them fail then the error secret is created.
```

Comment:
The 'failed list' is per TaskRun. There is no global 'Failed List' (which would really require CRDs to manage state). Even if a host is considered failed for one task subsequent tasks will still attempt to use it, so a minor glitch will not take it down completely. 

This could definitely be improved upon.